src/votekit/ballot_generator.py

```diff
@@ -0,0 +1,558 @@
+                cand_i = ranking[i]
+                greater_cand_support = cand_support_dict[cand_i]
+                for j in range(i, len(ranking)):
+                    cand_j = ranking[j]
+                    cand_support = cand_support_dict[cand_j]
+                    prob *= greater_cand_support / (greater_cand_support + cand_support)
+            ranking_to_prob[ranking] = prob
+        return ranking_to_prob
+
+    def generate_profile(self) -> PreferenceProfile:
+
+        permutations = list(it.permutations(self.candidates, self.ballot_length))
+        ballot_pool = []
+
+        for bloc in self.bloc_voter_prop.keys():
+            num_ballots = int(self.number_of_ballots * self.bloc_voter_prop[bloc])
+            pref_interval_dict = self.pref_interval_by_bloc[bloc]
+
+            ranking_to_prob = self._calc_prob(
+                permutations=permutations, cand_support_dict=pref_interval_dict
+            )
+
+            indices = range(len(ranking_to_prob))
+            prob_distrib = list(ranking_to_prob.values())
+            prob_distrib = [float(p) / sum(prob_distrib) for p in prob_distrib]
+
+            ballots_indices = choice(
+                indices,
+                num_ballots,
+                p=prob_distrib,
+                replace=True,
+            )
+
+            rankings = list(ranking_to_prob.keys())
+            ballots = [rankings[i] for i in ballots_indices]
+
+            ballot_pool = ballot_pool + ballots
+
+        pp = self.ballot_pool_to_profile(
+            ballot_pool=ballot_pool, candidates=self.candidates
+        )
+        return pp
+
+
+class AlternatingCrossover(BallotGenerator):
+    def __init__(
+        self,
+        slate_to_candidate: dict,
+        pref_interval_by_bloc: dict,
+        bloc_voter_prop: dict,
+        bloc_crossover_rate: dict,
+        **data
+    ):
+        """
+        Alternating Crossover Ballot Generation Model
+
+        Args:
+            slate_to_candidate (dict): a mapping of slate to candidates
+            (ex. {race: [candidate]})
+            pref_interval_by_bloc (dict): a mapping of bloc to preference interval
+            (ex. {race: {candidate : interval length}})
+            bloc_voter_prop (dict): a mapping of the percentage of total voters per bloc
+            (ex. {race: 0.5})
+            bloc_crossover_rate (dict): a mapping of percentage of crossover voters per bloc
+            (ex. {race: {other_race: 0.5}})
+        """
+        # Call the parent class's __init__ method to handle common parameters
+        super().__init__(**data)
+
+        # Assign additional parameters specific to AC
+        self.slate_to_candidate = slate_to_candidate
+        self.pref_interval_by_bloc = pref_interval_by_bloc
+        self.bloc_voter_prop = bloc_voter_prop
+        self.bloc_crossover_rate = bloc_crossover_rate
+
+        # assert sum(bloc_crossover_rate.values()) == 1
+        assert sum(bloc_voter_prop.values()) == 1
```

Comment:
note: assertions can be disabled in production (https://realpython.com/python-assert-statement/#using-assert-for-data-processing-and-validation)

probably best to write something like:

```
if sum(...) != 1:
    raise ValueError("message here")
```